bolt/src/main/java/com/slack/api/bolt/App.java

```diff
@@ -314,26 +317,60 @@ private void initOAuthServicesIfNecessary() {
+     * Get the Slack URL for beginning the OAuth flow, including the query
+     * params necessary to identify this application to Slack.
+     *
+     * Appends the optional `redirect_uri` query param based on the provided
+     * AppConfig to ensure that the correct OAuth redirect URI is selected in
+     * cases where a Slack application may have multiple redirect URIs
+     * associated with it.
+     *
+     * @param state The OAuth state param
+     * @return The Slack URL to redirect users to for beginning the OAuth flow
+     */
     public String getOauthInstallationUrl(String state) {
         AppConfig config = config();
+
         if (config.getClientId() == null || config.getScope() == null || state == null) {
             return null;
+        }
+
+        String scope = config.getScope() == null ? "" : config.getScope();
+        String redirectUriParam = redirectUriQueryParam(appConfig);
+
+        if (config.isClassicAppPermissionsEnabled()) {
+            // https://api.slack.com/authentication/migration
+            return "https://slack.com/oauth/authorize" +
+                    "?client_id=" + config.getClientId() +
+                    "&scope=" + scope +
+                    "&state=" + state +
+                    redirectUriParam;
         } else {
-            String scope = config.getScope() == null ? "" : config.getScope();
-            if (config.isClassicAppPermissionsEnabled()) {
-                // https://api.slack.com/authentication/migration
-                return "https://slack.com/oauth/authorize" +
-                        "?client_id=" + config.getClientId() +
-                        "&scope=" + scope +
-                        "&state=" + state;
-            } else {
-                String userScope = config.getUserScope() == null ? "" : config.getUserScope();
-                return "https://slack.com/oauth/v2/authorize" +
-                        "?client_id=" + config.getClientId() +
-                        "&scope=" + scope +
-                        "&user_scope=" + userScope +
-                        "&state=" + state;
-            }
+            String userScope = config.getUserScope() == null ? "" : config.getUserScope();
+            return "https://slack.com/oauth/v2/authorize" +
+                    "?client_id=" + config.getClientId() +
+                    "&scope=" + scope +
+                    "&user_scope=" + userScope +
+                    "&state=" + state +
+                    redirectUriParam;
+        }
+    }
+
+    private String redirectUriQueryParam(AppConfig appConfig) {
+        if (appConfig.getRedirectUri() == null) {
+            return "";
+        }
+
+        try {
+            String urlEncodedRedirectUri = URLEncoder.encode(
+                appConfig.getRedirectUri(),
+                StandardCharsets.UTF_8.name()
+            );
+
+            return String.format("&redirect_uri=%s", urlEncodedRedirectUri);
+        } catch (UnsupportedEncodingException e) {
```

Comment:
`URLEncoder.encode` is virtually guaranteed to not throw because I am using a built-in encoding charset for URL-encoding the value for the `redirect_uri` query parameter. However, I'm open to suggestions for clarifying this a bit, or at least making this exception visible through logging. Throwing this exception seems inappropriate from this call site because it would require exception handling all the way up the stack in away that would change the public API of `App`.

Reply:
> suggestions for clarifying this a bit, or at least making this exception visible through logging.  

As this is a well-known thing in Java and the exception is never thrown for sure, I don't see any issues with the current code.